File: test-fixtures/sample.rb
require 'json'
require 'digest'
require 'securerandom'
require 'set'
require 'monitor'

module UserRoles
  ADMIN = :admin
  USER = :user
  VIEWER = :viewer
end

class UserConfig
  attr_accessor :name, :email, :role, :settings

  def initialize(name:, email:, role: UserRoles::VIEWER, settings: {})
    @name = name
    @email = email
    @role = role
    @settings = settings
    @created_at = Time.now
  end

  def validate
    return false if @name.nil? || @name.length < 2
    return false unless @email.include?('@')
    return false unless [UserRoles::ADMIN, UserRoles::USER, UserRoles::VIEWER].include?(@role)
    true
  end

  def to_h
    { name: @name, email: @email, role: @role, created_at: @created_at.iso8601 }
  end
end

class ApiResponse
  attr_reader :status, :data, :message, :timestamp

  def initialize(status:, data: nil, message:)
    @status = status
    @data = data
    @message = message
    @timestamp = (Time.now.to_f * 1000).to_i
  end

  def ok?
    @status >= 200 && @status < 300
  end

  def to_h
    { status: @status, data: @data, message: @message, timestamp: @timestamp }
  end
end

class BaseService
  def initialize(base_url, timeout = 5)
    @base_url = base_url
    @timeout = timeout
  end

  def log(method, path)
    puts "[#{method}] #{@base_url}#{path}"
  end
end

class UserService < BaseService
  include MonitorMixin

  def initialize
    super('https://api.example.com/users')
    @users = {}
    @listeners = []
    super() # init monitor
  end

  def get_user(id)
    synchronize do
      user = @users[id]
      return ApiResponse.new(status: 404, message: 'Not found') unless user
      @listeners.each { |l| l.call(:fetched, user) }
      ApiResponse.new(status: 200, data: user.to_h, message: 'OK')
    end
  end

  def create_user(config)
    return ApiResponse.new(status: 400, message: 'Invalid config') unless config.validate
    id = SecureRandom.hex(16)
    synchronize do
      @users[id] = config
    end
    @listeners.each { |l| l.call(:created, config) }
    ApiResponse.new(status: 201, data: config.to_h, message: 'Created')
  end

  def update_user(id, updates)
    synchronize do
      user = @users[id]
      return ApiResponse.new(status: 404, message: 'Not found') unless user
      user.name = updates[:name] if updates.key?(:name)
      user.email = updates[:email] if updates.key?(:email)
      ApiResponse.new(status: 200, data: user.to_h, message: 'Updated')
    end
  end

  def delete_user(id)
    synchronize do
      return ApiResponse.new(status: 404, message: 'Not found') unless @users.key?(id)
      @users.delete(id)
    end
    ApiResponse.new(status: 200, message: 'Deleted')
  end

  def list_users
    synchronize do
      @users.values.map(&:to_h)
    end
  end

  def on_event(&block)
    @listeners << block
  end
end

module Utils
  def self.calculate_hash(content)
    Digest::SHA256.hexdigest(content)
  end

  def self.validate_email(email)
    email&.include?('@') && email&.include?('.')
  end

  def self.paginate(items, page, per_page)
    start_idx = (page - 1) * per_page
    end_idx = start_idx + per_page - 1
    pages = (items.length.to_f / per_page).ceil
    { items: items[start_idx..end_idx] || [], total: items.length, pages: pages }
  end

  def self.read_config(path)
    JSON.parse(File.read(path))
  rescue JSON::ParserError, Errno::ENOENT
    {}
  end

  def self.write_config(path, config)
    File.write(path, JSON.pretty_generate(config))
  end

  def self.retry(times, delay, &block)
    attempts = 0
    begin
      attempts += 1
      block.call
    rescue StandardError => e
      raise e if attempts >= times
      sleep(delay)
      retry
    end
  end
end

File: test-fixtures/sample.py
import os
import sys
import json
import hashlib
import asyncio
import logging
from typing import Optional, Dict, List, Any, Tuple
from datetime import datetime
from enum import Enum
from abc import ABC, abstractmethod

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class HttpMethod(Enum):
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"
    PATCH = "PATCH"


class ApiResponse:
    def __init__(self, status: int, data: Any, message: str):
        self.status = status
        self.data = data
        self.message = message
        self.timestamp = int(datetime.now().timestamp())

    def to_dict(self) -> Dict:
        return {
            "status": self.status,
            "data": self.data,
            "message": self.message,
            "timestamp": self.timestamp,
        }


class UserConfig:
    def __init__(self, name: str, email: str, role: str = "viewer"):
        self.name = name
        self.email = email
        self.role = role
        self.created_at = datetime.now()

    def validate(self) -> bool:
        if not self.name or len(self.name) < 2:
            return False
        if "@" not in self.email:
            return False
        if self.role not in ("admin", "user", "viewer"):
            return False
        return True

    def to_dict(self) -> Dict:
        return {
            "name": self.name,
            "email": self.email,
            "role": self.role,
            "created_at": self.created_at.isoformat(),
        }


class BaseService(ABC):
    def __init__(self, base_url: str, timeout: int = 5000):
        self.base_url = base_url
        self.timeout = timeout

    @abstractmethod
    async def handle(self, data: Any) -> ApiResponse:
        pass

    def _log(self, method: str, path: str) -> None:
        logger.info(f"[{method}] {path}")


class UserService(BaseService):
    def __init__(self):
        super().__init__("https://api.example.com/users")
        self._users: Dict[str, UserConfig] = {}

    async def handle(self, data: Any) -> ApiResponse:
        return await self.get_user(str(data))

    async def get_user(self, user_id: str) -> ApiResponse:
        user = self._users.get(user_id)
        if not user:
            return ApiResponse(404, None, "Not found")
        return ApiResponse(200, user.to_dict(), "OK")

    async def create_user(self, config: UserConfig) -> ApiResponse:
        if not config.validate():
            return ApiResponse(400, None, "Invalid config")
        user_id = hashlib.sha256(os.urandom(32)).hexdigest()
        self._users[user_id] = config
        return ApiResponse(201, config.to_dict(), "Created")

    async def update_user(self, user_id: str, updates: Dict) -> ApiResponse:
        existing = self._users.get(user_id)
        if not existing:
            return ApiResponse(404, None, "Not found")
        for key, value in updates.items():
            setattr(existing, key, value)
        return ApiResponse(200, existing.to_dict(), "Updated")

    async def delete_user(self, user_id: str) -> ApiResponse:
        if user_id not in self._users:
            return ApiResponse(404, None, "Not found")
        del self._users[user_id]
        return ApiResponse(200, None, "Deleted")

    def list_users(self) -> List[UserConfig]:
        return list(self._users.values())


class SearchEngine:
    def __init__(self, index_dir: str):
        self.index_dir = index_dir
        self.documents: Dict[str, str] = {}

    def index_file(self, filepath: str) -> None:
        if not os.path.exists(filepath):
            return
        with open(filepath, "r") as f:
            content = f.read()
        doc_id = hashlib.md5(filepath.encode()).hexdigest()
        self.documents[doc_id] = content

    def search(self, query: str) -> List[Tuple[str, float]]:
        results = []
        query = query.lower()
        for doc_id, content in self.documents.items():
            if query in content.lower():
                score = content.lower().count(query) / len(content)
                results.append((doc_id, score))
        return sorted(results, key=lambda x: x[1], reverse=True)[:10]


def calculate_hash(content: str) -> str:
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def validate_email(email: str) -> bool:
    import re
    return bool(re.match(r"^[^\s@]+@[^\s@]+\.[^\s@]+$", email))


def paginate(items: List[Any], page: int, per_page: int) -> Dict:
    start = (page - 1) * per_page
    end = start + per_page
    return {
        "items": items[start:end],
        "total": len(items),
        "pages": -(-len(items) // per_page),
    }


async def retry(fn, retries: int, delay: float):
    for i in range(retries):
        try:
            return await fn()
        except Exception as e:
            if i == retries - 1:
                raise
            await asyncio.sleep(delay)
